### DISCUSSION TITLE
How can I check to see if a contract was already deployed and if not deploy a new one.

### DISCUSSION THREAD

**Discussion by LanceAddison** [Cyfrin/foundry-full-course-cu#2406] [Q&A] [answered]
I would like to first check if a contract has been deployed already and if not a new one should be deployed. I call this in a deploy script and am using DevOpsTools to call get_most_recent_deployment. If the function returns the zero address then the if statement in the code below creates a new contract. The problem is that the get_most_recent_deployment function reverts if it doesn't find a deployed contract. This means that it doesn't return the zero address so my if statement and rest of my deploy script is not completed. Additionally when I go to run a test it doesn't even get to the test because it sees there was a revert from the get_most_recent_deplyment function and figurest the test  should fail. Does anyone have an idea how I can get this functionality to work. 

Deploy contract:
'''
contract DeployMintiVerseMarket is Script {
    ArtShift artShift;
    MintiVerseMarket mintiVerse;
    HelperConfig helper;

    function run() external returns (ArtShift, MintiVerseMarket, HelperConfig) {
        address artShiftAddress = DevOpsTools.get_most_recent_deployment("ArtShift", block.chainid);

        if (artShiftAddress == address(0)) {
            DeployArtShift deployer = new DeployArtShift();
            (artShift, helper) = deployer.run();
        } else {
            artShift = ArtShift(artShiftAddress);
            helper = HelperConfig(DevOpsTools.get_most_recent_deployment("HelperConfig", block.chainid));
        }

        (,,,,, uint256 deployerKey) = helper.activeNetworkConfig();

        vm.startBroadcast(deployerKey);
        mintiVerse = new MintiVerseMarket(address(artShift));
        vm.stopBroadcast();

        return (artShift, mintiVerse, helper);
    }
}
'''

Here is the error that I received which is the revert returned by the get_most_recent_deployment function:
```
[FAIL. Reason: setup failed: revert: No contract deployed] setUp() (gas: 0)
Suite result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 114.45ms (0.00ns CPU time)
```

**Comment by EngrPips** [comment, marked as answer]
Hello @LanceAddison, The `DevOpsTools` was created to always get an already deployed instance of a contract and work with it, so before you use the `DevOpsTools` in a script, first ensure that you have run a deployment script to deploy an instance of the contract you want to work with. If you want more flexibility, as you have said, you can open the `DevOpsTools` and study how it operates and edit its codebase to serve your need, or you can create your own `DevOpsTools` from scratch, which will search for a deployed instance of a contract and if it didn't see one, it will return address(0).

**Reply by LanceAddison** [reply]
Ok thanks for the answer. I figured I would have to modify it. Just wasn't sure if there was another way.

**Reply by EngrPips** [reply]
You are welcome, Friend.
